### PR TITLE
Hongyi sql refactor

### PR DESCRIPTION
**Pull request by hongyishi**
# Description

A new interface for SQL database indexing. SQLTableIndex only operates on pre-existing SQL databases. Functionally it should be the same as the existing SQLStructStoreIndex.
The main difference is that the actual index only stores the SQL database reference and table info into the passed in VectorStoreIndex. It does not interface with Nodes on its own.

There is a new SQLTableRetriever that can retrieve based on specified table names or llm prediction.
A new NLSQLTableQueryEngine takes in the SQLTableIndex. The main difference is its _get_table_context no longer uses _sql_context_container.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [ ] Added new notebook (that tests end-to-end) (I actually modified the existing Airbyte workflow but I can separate it out or leave it as the new version.)

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/indices/struct_store/sql_table.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            table_text = f"Schema of table {table_name}:\n" f"{table_desc}\n"
+            docs.append(Document(table_text, extra_info={"table_name": table_name}))
+        for table_name, table_context in self._extra_context_dict.items():
+            docs.append(
+                Document(
+                    f"{table_name}: {table_context}",
+                    extra_info={"table_name": table_name},
+                )
+            )
+        nodes = self._service_context.node_parser.get_nodes_from_documents(docs)
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/sql_table.py:83]: hm not sure there's a reason to do this, you can just construct Node objects above directly

File: llama_index/indices/struct_store/retriever.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    def _retrieve(
+        self,
+        query_bundle: QueryBundle,
+    ) -> List[NodeWithScore]:
+        """Retrieve nodes."""
+        vector_store_index = self._index.vector_store_index
+
+        # Specific table case. Use initialized tables if specified.
+        if self._tables:
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/retriever.py:41]: i'm not actually sure what this is supposed to do. if `self._tables` is not specified, then you do retrieval from the existing index based on the query, which is fair. If it is specified...you feed the schema in and get back a similar document? I'm confused on the use case

File: llama_index/indices/struct_store/retriever.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            vector_index_retriever = vector_store_index.as_retriever(similarity_top_k=1)
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/retriever.py:42]: why is similarity_top_k=1 hardcoded

Comment (jerryjliu) [reply llama_index/indices/struct_store/retriever.py:42]: also see comments below, not sure why this needs to be vector store specific

File: llama_index/indices/struct_store/sql_table.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        self._build_index_from_tables()
+
+    def _build_index_from_tables(self) -> None:
+        """Build index from sql tables."""
+        docs = []
+        for table_name in self.sql_database.get_usable_table_names():
+            table_desc = self.sql_database.get_single_table_info(table_name)
+            table_text = f"Schema of table {table_name}:\n" f"{table_desc}\n"
+            docs.append(Document(table_text, extra_info={"table_name": table_name}))
+        for table_name, table_context in self._extra_context_dict.items():
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/sql_table.py:76]: so you add table_name and table_context as additional Documents? Could you explain the rationale for that?

File: llama_index/indices/struct_store/sql_query.py [hunk 5]
```diff
@@ -203,6 +206,134 @@ async def _aquery(self, query_bundle: QueryBundle) -> Response:
...
+        return response
+
+    async def _aquery(self, query_bundle: QueryBundle) -> Response:
+        return self._query(query_bundle)
+
+
+class NLSQLTableQueryEngine(BaseQueryEngine):
+    def __init__(
+        self,
+        index: SQLTableIndex,
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/sql_query.py:234]: i wonder if we should create two NL query engines. 
The first doesn't take in an index at all, just the sql_database. and _get_table_context just dumps the table schema from the sql database class.

The second actually does retrieval from the table index.

**Review by jerryjliu — COMMENTED** [review]
Some high-level comments.

Also once we agree on overall design would be good to edit the existing sql_guide.md and existing SQL notebooks. If we agree on design we should swap out all instances of old SQL flow for new SQL flow in docs/notebooks. 

Also once we align on design we should add tests

File: llama_index/indices/struct_store/retriever.py (hunk 2 above)
Comment (hongyishi) [reply llama_index/indices/struct_store/retriever.py:41]: Hmm yeah so the problem is we lose reference to the individual nodes once they are written into the store, so I need to do a query on the store to get the right schema back.

File: llama_index/indices/struct_store/sql_table.py (hunk 1 above)
Comment (hongyishi) [reply llama_index/indices/struct_store/sql_table.py:83]: Fixed

File: llama_index/indices/struct_store/sql_table.py (hunk 4 above)
Comment (hongyishi) [reply llama_index/indices/struct_store/sql_table.py:76]: This is for adding additional context info beyond the default table schema

File: llama_index/indices/struct_store/retriever.py (hunk 3 above)
Comment (hongyishi) [reply llama_index/indices/struct_store/retriever.py:42]: Doesn't have to be a vector store actually.

File: llama_index/indices/struct_store/sql_query.py (hunk 5 above)
Comment (hongyishi) [reply llama_index/indices/struct_store/sql_query.py:234]: Hmm the retrieval is actually happening in SQLTableRetriever though? The NLQueryEngine mainly orchestrates the NL to SQL prediction.

Comment (jerryjliu) [reply llama_index/indices/struct_store/sql_query.py:234]: yeah but if you pass in sql_database directly then you can just do `sql_database.table_info` and dump that into the prompt. easier ergonomics for user if they're not worreid about retrieval

File: llama_index/indices/struct_store/retriever.py (hunk 3 above)
Comment (jerryjliu) [reply llama_index/indices/struct_store/retriever.py]: why is that the case?

Comment (hongyishi) [reply llama_index/indices/struct_store/retriever.py]: Ah you mean including extra context info. Hmm..

Comment (hongyishi) [reply llama_index/indices/struct_store/retriever.py]: Fixed

File: llama_index/indices/struct_store/sql_query.py [hunk 6]
```diff
@@ -203,6 +206,150 @@ async def _aquery(self, query_bundle: QueryBundle) -> Response:
...
+        """Answer a query."""
+        # NOTE: override query method in order to fetch the right results.
+        # NOTE: since the query_str is a SQL query, it doesn't make sense
+        # to use ResponseBuilder anywhere.
+        response_str, extra_info = self._sql_database.run_sql(query_bundle.query_str)
+        response = Response(response=response_str, extra_info=extra_info)
+        return response
+
+    async def _aquery(self, query_bundle: QueryBundle) -> Response:
+        return self._query(query_bundle)
```
Comment (Disiok) [inline llama_index/indices/struct_store/sql_query.py:228]: Hmm so for this module we don't even use the LLM at all? Neither text to sql query, nor sql response to text?

Comment (jerryjliu) [reply llama_index/indices/struct_store/sql_query.py:228]: yeah i think i mentioned this earlier cc @hongyishi but we probably don't need this (this wrapper around the raw sql_database) doesn't really do much

File: llama_index/indices/struct_store/sql_query.py [hunk 7]
```diff
@@ -203,6 +206,150 @@ async def _aquery(self, query_bundle: QueryBundle) -> Response:
...
+        logger.debug(f"> Predicted SQL query: {sql_query_str}")
+
+        response_str, extra_info = self._sql_database.run_sql(sql_query_str)
+        extra_info["sql_query"] = sql_query_str
+        response = Response(response=response_str, extra_info=extra_info)
+        return response
+
+
 # legacy
 GPTNLStructStoreQueryEngine = NLStructStoreQueryEngine
```
Comment (jerryjliu) [inline llama_index/indices/struct_store/sql_query.py:354]: might be good to mark `NLStructStoreQueryEngine` and `SQLStructStoreQueryEngine` as deprecated in comments/docstring

**Review by jerryjliu — COMMENTED** [review]
@hongyishi as we're waiting for the object index stuff, i'm down to just land `NLSQLTableQueryEngine` as its own PR if you wanted to split that out.

will get to object index tonight

**Review by jerryjliu — COMMENTED** [review]
nice - one final piece is adding a unit test (see tests/indices/struct_store/test_sql_query.py as an example)

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**